### ISSUE TITLE
kloud: provider improvements

### ISSUE THREAD

**Issue by rjeczalik** [koding/koding#9266]
A TODO list of improvements for existing kloud providers:

~~\- Improve credential validation; for now `google` and `digitalocean` (*Credential).Valid() returns nil error for a Credential zero-value:~~

```
$ cat /Volumes/rjk.io/aws-dev.json | ./kd c create -f - -team foobar
Creating credential... 
Error communicating with Koding: stack: explicit provider name is required for this credential; matches [aws digitalocean google] (c140dedfe24ab438)
```
- improve generated schemas, adding enums for `azure` (mostly storage-related), `google` and `digitalocean`; /cc'ing @gokmen - I remember there was a feature request to extend schema to  include also `instance type -> (cpu count, memory size)` mapping per provider.

_(more coming)_

**Comment by ppknap** [comment]
@rjeczalik It looks like you need to improve your current wip branch `stack-create`:

I assume that [here](https://github.com/koding/koding/blob/1c98c5aab4161100f8a84c1aa190ec19cb6f1fd3/go/src/koding/kites/kloud/stack/credential.go#L258) you wanted to add:

```go
cred := p.NewCredential()
```

and [here](https://github.com/koding/koding/blob/1c98c5aab4161100f8a84c1aa190ec19cb6f1fd3/go/src/koding/kites/kloud/stack/credential.go#L265-L267) you were going to add provider name into `matching` slice when `err == nil`:

```go
if err := v.Valid(); err == nil {
	matching = append(matching, name)
}
```

**Comment by gokmen** [comment]
@ppknap got this while bootstrap on google provider; 

```
[18:02:50] An error occurred: apply failed with code: 1, output: e:Error loading config: 
Error loading /.../go/data/terraformer/kodingdev-vms-camel-131e94842ee37081a3860285485567c1/main.tf.json: 
position -: resource must be followed by exactly two strings, a type and a name
```

**Comment by rjeczalik** [comment]
@ppknap It's most likely a tf bug, you can apply a workaround like:

https://github.com/koding/koding/blob/master/go/src/koding/kites/kloud/stack/provider/builder.go#L537-L543

**Comment by ppknap** [comment]
@gokmen I made a workaround for this issue. However, there is another bug on hashicorp side. Their hcl parser panics on interface values that store structures. I'm gonna to send them PR with the fix. Until then you will see:

![image](https://cloud.githubusercontent.com/assets/5021246/19417219/5dbfc46e-93a7-11e6-893a-fcbee6a88323.png)

**Comment by gokmen** [comment]
@ppknap we need to fix these issues before enabling gce provider. cc/ @rjeczalik

**Comment by gokmen** [comment]
@rjeczalik there is an issue for DO provider which is not updating `droplet_id` in `JMachine.meta` after build complete which is causing `start/stop` not working properly after build.

**Comment by rjeczalik** [comment]
Two issues with DO:

- droplet ID is missing in machine's medata
- ssh keys during bootstrap should be a unique resource

**Comment by gokmen** [comment]
@ppknap for GCE, it's unable to turn off the vm, getting timeout error after 10min. passed

**Comment by ppknap** [comment]
@gokmen will look at this today morning(it's 2am here). I'm also going to add unique ID for each machine ID since it's not possible to reinitialize the stack because of name duplication caused by async nature of re-initialization process.

**Comment by ppknap** [comment]
Currently, `terraform plan/apply` is not able to obtain GCE instance's disc size if it's not set explicitly in template. However, due to old terraform lib, we always try to get google image from image family using GCP API. This api returns [Image](https://godoc.org/google.golang.org/api/compute/v1#Image) object that also contains image size. What about setting disc size in terraform template based on these data. This should improve UX by correctly reporting disc size. @gokmen @rjeczalik wdyt ?

**Comment by gokmen** [comment]
it looks ok to me since it will be the correct size if it's not defined explicitly.

**Comment by gokmen** [comment]
@ppknap with the latest changes for GCP I'm getting following error when I try to `start` a `stopped` machine. which is correct based on the data on `JMachine.status.state`

![image](https://cloud.githubusercontent.com/assets/42368/19836772/5e9d61e0-9eb3-11e6-8077-fb56a4d3b7a5.png)

I think this needs to be solved by updating status as `Stopped` when we got `Terminated` after a `stop` request.

As a side note; if I change `status.state` to `Stopped` manually it works well.

cc/ @rjeczalik

**Comment by ppknap** [comment]
@gokmen this is not related with latest change. It seems that we use machine.Info to update JMachine state and my workaround for google API will not work. I'll send a "fix" in few minutes.

**Comment by ppknap** [comment]
@gokmen https://github.com/koding/koding/pull/9423

**Comment by gokmen** [comment]
@rjeczalik in Azure provider I'm getting following error when I request bootstrap;

```
[17:57:40] Parsing template for credential requirements...
[17:57:40] Following credentials are required:
[17:57:40] - [
	"azure"
]
[17:57:40] Parsing template for additional requirements...
[17:57:41] Template content saved.
[17:57:41] Setting up custom variables...
[17:57:41] Custom variables are set.
[17:57:41] Checking provided credentials...
[17:57:41] Verifying credentials...
[17:57:41] Bootstrap check initiated for credentials...
[17:57:42] Bootstrap required, initiating to bootstrap...
[17:57:44] An error occurred: apply failed with code: 1, output: o:There are warnings and/or errors related to your configuration. Please
fix these before continuing.

e:Errors:

e:  * provider.azure: error parsing publish_settings as XML: EOF
```

**Comment by gokmen** [comment]
@rjeczalik I've my credential `publish_settings` as in `base64` format and it passes validation step from `kloud` but on `bootstrap` I'm getting the error above. also for reference my stack template;

```
provider:
  azure:
    publish_settings: '${var.azure_publish_settings}'
    subscription_id: '${var.azure_subscription_id}'

resource:
  azure_instance:
    # this is the name of your VM
    azure-instance:
      # select your instance size here: eg. Basic_A1
      size: Basic_A1
      # base image for your instance
      image: 'Ubuntu Server 14.04 LTS'
      # on custom_data section we will write bash and configure our VM
      custom_data: |-
        # let's create a file on your root folder:
        echo "hello world!" >> /helloworld.txt
```

and my `JCredential.meta` content (`publish_settings` trimmed)

```
{
  "publish_settings": "PD94bWwgdmVyc2l....HVibGlzaERhdGE+",
  "subscription_id": "aa3ec1eb-ef8c-4010-b2f9-11cf61156cb4",
  "location": "East US 2",
  "storage": "Standard_LRS",
  "identifier": "09216c619c8d6b0caec746c5b7535c26"
}
```

and verified that it's a valid XML as in `publish_settings`

**Comment by gokmen** [comment]
@rjeczalik you can try it on `providers` branch https://github.com/koding/koding/pull/9319 fyi.

**Comment by rjeczalik** [comment]
The reason for:

```
* provider.azure: error parsing publish_settings as XML: EOF
```

Was publish_settings, which were not base64-encoded when read from sneaker. I've dropped the base64 encoding of `publish_settings` field, and now it works. I've also made a number of other improvements that I could spot.

With current ui I was not able to get past `Initialize stack`, after clicking the button stack disappeared.  Moreover each time I click save the bootstrap is being creating all over again - let me know if that's something I should look into.
